pip/qsharp/_qsharp.py

```diff
@@ -93,23 +93,60 @@ def eval_file(path: str) -> Any:
     return eval(f.read())
 
 
-def run(entry_expr: str, shots: int) -> Any:
+class ShotResult(TypedDict):
+    """
+    A single result of a shot.
+    """
+
+    events: List[Output]
+    result: Any
+
+
+def run(
+    entry_expr: str,
+    shots: int,
+    on_result: Optional[Callable[[ShotResult], None]] = None,
+    save_output: bool = False,
+) -> Any:
     """
     Runs the given Q# expression for the given number of shots.
     Each shot uses an independent instance of the simulator.
 
     :param entry_expr: The entry expression.
     :param shots: The number of shots to run.
+    :param on_result: A callback function that will be called with each result.
+    :param save_output: If true, the output of each shot will be saved. If false, they will be printed.
 
-    :returns values: A list of results or runtime errors.
+    :returns values: A list of results or runtime errors. If `save_output` is true,
+    a Dict is returned with a "results" property containing the list of results,
+    and an "events" property containing a list of lists of events for each result.
 
```

Comment:
Should `"events"` really be called `"outputs"` (my preference)? Or should `save_outputs` be called `save_events`? Events are just output events, so from a user perspective it reads like we're calling the same thing two different names.

Reply:
I didn't like this either. In the compiler today we call them events (https://github.com/microsoft/qsharp/blob/main/npm/src/compiler/common.ts#L82), and that's what is surfaced in the wasm interface too. But I see they got called an 'Output' when the Python class was implemented.

Peronally I'd rather keep them consistent across APIs with 'events', so can make that change.

Reply:
> Events are just output events

There's also no guarantee this will always be true. We may at some point have tracing or profiling events we want to capture, for example.

Reply:
"events" sounds good to me, thanks

Reply:
I disagree slightly, but not enough to block. The reason why the line you point to Bill uses the term "events" is that it is for messages and dump machine, not for the final result of execution. These are two different things that we are combining for the first time here.

Reply:
This is just for Message and DumpMachine. We save the results of execution regardless. We only save the events (Message and DumpMachine currently) in this field (and only if save_events is True).